### ISSUE TITLE
[New feature request] Early exit network : requires join element.

### ISSUE THREAD

**Issue by gichan-jang** [nnstreamer/nnstreamer#2814]
As networks become more larger and larger, this is limited in real-time processing and low power programs.  
In order to alleviate these problems, If confidence reaches a certain level, just use the results without processing the data anymore.  
The concept of pipeline to implement this is as follows:  

![Early_exit_pipeline2](https://user-images.githubusercontent.com/56856496/96998533-8fea1280-156e-11eb-8fd0-f41369d87e0e.png)

In tensor(s) Zab, a and b mean the model passed. Tensor Z12 and tensor Z13 have the same tensor(s) config.  
The collector sends the tensor(s) to the sink as soon as it arrives.  

There are tensor_if example that utilizes model output confidence:  
https://github.com/nnstreamer/nnstreamer-example/tree/master/bash_script/example_tensorif

To impelement above the concept, The collector element is needed.

**Comment by dongju-chae** [comment]
How about extend `tensor_if` to handle data from multiple sink pads?

**Comment by gichan-jang** [comment]
> How about extend `tensor_if` to handle data from multiple sink pads?

It is also possible to implement by extending `tensor_if`.
`tensor_if` is determining output based on tensor data with a given condition, but the new element is irrelevant to the condition. 
So I think it would be better to add a new element.

**Comment by myungjoo** [comment]
@jijoongmoon Please invite Cambridge Early-Exit researchers. We are drafting pipelines for early exit, which will allow them not to implement/hack pytorch/tensorflow, and allow them to focus on the algorithm itself.

**Comment by myungjoo** [comment]
```tensor_selector``` is an ```input-selector``` that selects any incoming sink-pad or a ```tensor-mux``` that has two sink-pads with the same caps running at "refresh mode" and the output (src-pad) is the pass-through of active input.

**Comment by myungjoo** [comment]
Anyway we may provide a "GST BIN" of ```{ tensor-if + tensor-filter + tensor-selector }``` as a syntactic suger.

**Comment by kparichay** [comment]
In the proposed design, each `early_exit` module represents a particular subsection of the model (this subsection would provide two outputs, the first output is the features Y from the backbone (model1) to be used by the classifier and the second is the output Z12 by the classifier (model2)). In the case of N confidence checks, this would require splitting a model into `2*N - 1` models to fit into the pipeline structure which isn't easy IMO (or loading the full model in all tensor_filter which is overkill). Another disadvantage is that in order to add a confidence check, the feature extractor needs to be further broken up into yet another model.

Is it possible to pass the output of a tensor_filter to itself conditional based on tensor_if, possibly with help of tensor_repo (Model1 output Y can be passed back to Model1 by tensor_If if not enough confidence, or passed forward - this would not require collector as there is just 1 output)? If yes, then the splitting of the feature extractor can be avoided. 
Assuming that the feature extractor model has been compiled to be able to run with multiple inputs/outputs, so the model can run with for different features using the input from the previous run of the model itself (this is possible in TensorFlow-Lite with choosing input/output index and TensorFlow with input/output name). In this case, each confidence check will have its own model for classification, and the feature extractor/backbone will be shared (reducing the total number of models to N+1). The input/output to be run can be chosen using from the metadata in the tensor or tensor_if can send a signal to tensor_filter. 
Advantage: Fewer splits of the original model, and adding a new confidence interval does not need splitting existing feature extractor but just choosing a different input/output index in it.

**Comment by gichan-jang** [comment]
![Early_exit_pipeline_feedback_loop](https://user-images.githubusercontent.com/56856496/97134135-9f947180-178f-11eb-80c6-6212734f24fa.png)

@kparichay 
Does this match what you suggested?
To implement this, the tensor filter seems to require a lot of changes.

**Comment by kparichay** [comment]
@gichan-jang This is pretty much what I had mind.

> To implement this, the tensor filter seems to require a lot of changes.

Yes, in this case, `tensor_filter` will require the following features:
- Support to select input/output indexes to run that running all of them/running just first one.
- Embedding the index information from `tensor_if` to `tensor_filter` as metadata in the buffer/sent via signal.
- Possibly some changes to `tensor_repo` (I am not familiar with tensor_repo, so I am not sure).

**Comment by myungjoo** [comment]
1. I'd recommend NOT to use tensor_repo for such purposes. You need a complex synchronization mechanism to ensure that you won't mix up different input frames here.
2. If you do not split up the backbone model into ```2 * N - 1``` submodels for Early Exit, you need to keep in mind that:
    - modifying tensorflow/tensorflow-lite/pytorch/caffe2 is not a tractable option for us.
    - you need to make sure that (at least) tf-lite and one/nnfw-runtime supports such behaviors.
    - need to avoid making tensor-filter core larger.
3. If you think having too many elements in a pipeline matters, GST-BIN with pre-defined elements might help (or something like decodebinm, which will arrange elements automagically based on the given properties/requirements.)

**Comment by kparichay** [comment]
1. I do not have any experience with tensor_repo. I will take your word for it. If your recommendation is that tensor_repo is not suitable for this task (given the complexity and difficulty of using it), then this idea can be scrapped.
2. In my understanding, Tensorflow-lite and TensorFlow (possibly others too) do support multiple inputs/multiple outputs with their infrastructure for the models. 
- modifying the framework is definitely not thought about when suggesting this style as it's out of our hands.
- For PyTorch/Caffe2, I will check and update here. For one/nnfw-runtime, I will confirm with the corresponding people.
- This might lead to a bigger tensor-filter core requiring the support for choosing multiple/selective inputs/outputs. However, this design reduces the number of tensor_filter elements and framework instances, reducing significant memory overhead, and possibly some performance overhead.
3. GST-BIN simplifies the pipeline from the readers perspective but still does not target the two major issues:
- The performance/memory overhead of these extra elements
- The overhead of splitting the models and adding new elements to the pipeline (note that in the proposed method, the pipeline will still need to be modified to specify the new inputs/outputs of the model).

**Comment by myungjoo** [comment]
> 1. I do not have any experience with tensor_repo. I will take your word for it. If your recommendation is that tensor_repo is not suitable for this task (given the complexity and difficulty of using it), then this idea can be scrapped.
> 2. In my understanding, Tensorflow-lite and TensorFlow (possibly others too) do support multiple inputs/multiple outputs with their infrastructure for the models.
> 
> * modifying the framework is definitely not thought about when suggesting this style as it's out of our hands.
> * For PyTorch/Caffe2, I will check and update here. For one/nnfw-runtime, I will confirm with the corresponding people.
> * This might lead to a bigger tensor-filter core requiring the support for choosing multiple/selective inputs/outputs. However, this design reduces the number of tensor_filter elements and framework instances, reducing significant memory overhead, and possibly some performance overhead.
> 

Tensorflow-lite/Tensorflow support multiple inputs/outputs, but I don't think it supports ```stop processing further if output-1 is supposed to be "enough"``` or ```pause processing when output-1 is ready and resume processing for output-2 if "continue" is signaled.```, which is why I'm inclined to separating models explicitly.

The Early-Exit researchers have been experimenting the system by adding hooks and replacing "no-ops" into PyTorch models with modifications of PyTorch. As you already know, we need mechanisms that do not need to touch PyTorch/Tensorflow.

**Comment by kparichay** [comment]
Although some frameworks do support it (with some extra work when making the model by making some variables as placeholders so that they can be fed in to allow running just some part of the models), it can't be done for all the frameworks.
So, maybe what @gichan-jang proposed is better that it supports all frameworks.

**Comment by gichan-jang** [comment]
The new element `join` which is called `tensor_collector` is merged. #2863 
I will upload a sample application to show the implementation of the early-exit network.

**Comment by gichan-jang** [comment]
Join Element required for early exit network is prepared and close this issue.
